File: 1 kyu/Square sums/solution.java
import java.util.*;
import java.util.stream.*;

public class SquareSums {
    
    final static private int TOP = 1200;
    final static private Set<Integer> SQ = IntStream.range(1, (int) Math.sqrt(2*TOP) + 1)
                                                    .map( n -> n*n ).boxed()
                                                    .collect(Collectors.toSet());
    
    
    public static List<Integer> buildUpTo(final int top) {
        
        List<Cnd> cnds = IntStream.range(1,top+1).mapToObj(Cnd::new).collect(Collectors.toList());
        for(final Cnd c: cnds) {
            SQ.stream().filter(  sq -> c.n<sq && sq-c.n<=top && 2*c.n!=sq )
                       .forEach( sq -> c.add( cnds.get(sq-c.n-1) ) );
        }
        Collections.sort(cnds);
        if (cnds.get(0).size()==1) cnds.removeIf(c->c.size()!=1);
        
        var out = new Stack<Integer>();
        return dfs(out, top, cnds) ? new ArrayList<>(out) : null;
    }
    
    
    private static boolean dfs(Stack<Integer> out, int top, List<Cnd> cnds) {
        
        if (out.size()==top) return true;
        if (cnds.isEmpty() || cnds.get(0).isEmpty() && out.size()+1!=top) return false;
        
        for (Cnd c: cnds) {
            List<Cnd> cuts = c.unLink();
            out.push(c.n);
            if (dfs(out,top, c.getCnds())) return true;
            out.pop();
            c.link(cuts);
        }
        return false;
    }
    
    
    //------------------------------------------------------------------
    
    private static class Cnd extends HashSet<Cnd> implements Comparable<Cnd> {
        private int n;
        
        private Cnd(int n){ super(); this.n=n; }
        
        private List<Cnd> getCnds(){ return this.stream().sorted().collect(Collectors.toList()); }
        
        private void link(List<Cnd> others) { 
            for(Cnd c: others) c.add(this); 
        }
        private List<Cnd> unLink() {
            List<Cnd> cuts = new ArrayList<>(this);
            for(Cnd c: cuts) c.remove(this);
            return cuts;
        }
        @Override public int compareTo(Cnd o) {
            return this.size()!=o.size() ? this.size()-o.size() : o.n-this.n; 
        }
        @Override public int     hashCode()       { return n; }
        @Override public boolean equals(Object o) { return o!=null && (o instanceof Cnd) && ((Cnd) o).n==this.n; }
        @Override public String  toString()       { 
            return String.format("Cnd(%d, {%s})", n, this.stream().map( c -> c.n+"" ).collect(Collectors.joining(","))); 
        }
    }
}
______________________________________________
import java.util.*;

public class SquareSums {
    
    public static List<Integer> buildUpTo(int n) {
        var graph = buildGraph(n);
        var candidates = new ArrayList<Integer>();
        for(int i=1;i<=n;i++) candidates.add(i);
        return DFS(graph, candidates, new ArrayList<Integer>(), n);
    }
  
    public static List<Set<Integer>> buildGraph(int n){
        var graph = new ArrayList<Set<Integer>>();
        var squares = new ArrayList<Integer>();
    
        for(int i=2;i*i<n*2;i++){
          squares.add(i*i);
        } 
        for(int i=0;i<=n;i++){
            graph.add(new HashSet<Integer>());
            for(int j:squares){
                if(i<j){
                    int diff= j-i;
                    if( diff == i ) continue;
                    if( diff <= n ) graph.get(i).add(diff);
                    else break;
                }
            }
        }   
        return graph;
    }
  
    public static List<Integer> DFS( List<Set<Integer>> graph, List<Integer> candidates, List<Integer> path, int n ){
        if( path.size() == n ) return path;
        candidates.sort((a, b) -> Integer.compare(graph.get(a).size(), graph.get(b).size()));
        for(var candidate: candidates){
            path.add(candidate);
            graph.get(candidate).forEach(e -> graph.get(e).remove(candidate));
            if(DFS(graph, new ArrayList<Integer>(graph.get(candidate)), path, n) != null) return path;
            path.remove(path.size()-1);
            graph.get(candidate).forEach(e -> graph.get(e).add(candidate));
        }
        return null;
    }
}
______________________________________
import java.util.*;
import java.util.function.Consumer;

public class SquareSums {
  public static final int MAX_N = 1600;
  private final int n;
  private final Consumer<List<Integer>> solutionProcessor;
  private final Integer[][] adjacent;
  private final Integer[] chain;
  private final boolean[] remaining;

  public SquareSums(int n, Consumer<List<Integer>> solutionProcessor) {
    if (n < 3)
      throw new IllegalArgumentException("n is too small");
    this.n = n;
    this.solutionProcessor = solutionProcessor;
    adjacent = new Integer[n + 1][];
    chain = new Integer[n];
    remaining = new boolean[n + 1];
    for (int i = 1; i <= n; i++) {
      adjacent[i] = connArray(i, n);
      chain[i - 1] = integerCache[i];
      remaining[i] = true;
    }
  }

  private static final Integer[] integerCache = new Integer[MAX_N + 1];
  static {
    for (int i = 0; i <= MAX_N; i++)
      integerCache[i] = i;
  }
  
  private static final boolean[][] connected = new boolean[MAX_N + 1][MAX_N + 1];
  static {
    int a = 1;
    while (true) {
      a++;
      int s = a * a;
      if (s > 2 * MAX_N - 1)
        break;
      for (int i = 1; i <= MAX_N; i++) {
        int j = s - i;
        if (j > 0 && j <= MAX_N && j != i)
          connected[i][j] = true;
      }
    }
  }

  private static Integer[] connArray(int i, int n) {
    List<Integer> connList = new ArrayList<>();
    boolean[] connRow = connected[i];
    for (int j = 1; j <= n; j++)
      if (connRow[j])
        connList.add(integerCache[j]);
    return connList.toArray(new Integer[connList.size()]);
  }

  private void swapChain(int i, int j) {
    if (i == j)
      return;
    Integer t = chain[i];
    chain[i] = chain[j];
    chain[j] = t;
  }

  private void solveNarrowed(int leftIndex, int rightIndex) {
    Integer left = integerCache[chain[leftIndex]];
    Integer right = integerCache[chain[rightIndex]];
    if (rightIndex == leftIndex + 2) {
      boolean[] lastConn = connected[chain[leftIndex + 1]];
      if (lastConn[left] && lastConn[right])
        solutionProcessor.accept(Arrays.asList(chain.clone()));
      return;
    }
    List<Integer> leftAdj = new ArrayList<>();
    List<Integer> rightAdj = new ArrayList<>();
    Integer needsLeft = null;
    Integer needsRight = null;
    for (int i = leftIndex + 1; i < rightIndex; i++) {
      int adjCount = 0;
      boolean aLeft = false;
      boolean aRight = false;
      Integer ii = integerCache[i];
      for (Integer j : adjacent[chain[i]])
        if (remaining[j]) {
          adjCount++;
          if (j == left) {
            aLeft = true;
            leftAdj.add(ii);
          } else if (j == right) {
            aRight = true;
            rightAdj.add(ii);
          }
        }
      if (adjCount < 2)
        return;
      if (adjCount == 2)
        if (aLeft)
          if (aRight || needsLeft != null)
            return;
          else
            needsLeft = ii;
        else if (aRight)
          if (needsRight != null)
            return;
          else
            needsRight = ii;
    }
    if (needsLeft != null)
      leftAdj = Collections.singletonList(needsLeft);
    int laCount = leftAdj.size();
    if (needsRight != null)
      rightAdj = Collections.singletonList(needsRight);
    int raCount = rightAdj.size();
    if (laCount == 0 || raCount == 0)
      return;
    if (laCount <= raCount) {
      remaining[left] = false;
      int nextLeftIndex = leftIndex + 1;
      for (int nextLeft : leftAdj) {
        swapChain(nextLeftIndex, nextLeft);
        solveNarrowed(nextLeftIndex, rightIndex);
        swapChain(nextLeftIndex, nextLeft);
      }
      remaining[left] = true;
    } else {
      remaining[right] = false;
      int prevRightIndex = rightIndex - 1;
      for (int prevRight : rightAdj) {
        swapChain(prevRight, prevRightIndex);
        solveNarrowed(leftIndex, prevRightIndex);
        swapChain(prevRight, prevRightIndex);
      }
      remaining[right] = true;
    }
  }

  // Solutions that are mirrors of each other are considered equal.
  // This also applies to solutionCount() and printAllSolutions().
  public void solveComplete() {
    int n1 = n - 1;
    for (int i = 0; i < n1; i++) {
      swapChain(0, i);
      for (int j = i + 1; j < n; j++) {
        swapChain(j, n1);
        solveNarrowed(0, n1);
        swapChain(j, n1);
      }
      swapChain(0, i);
    }
  }

  @SuppressWarnings("serial")
  private static class BreakSearchException extends RuntimeException {
  }

  public static List<Integer> firstSolution(int n) {
    @SuppressWarnings("unchecked")
    List<Integer>[] solutions = new List[1];
    try {
      new SquareSums(n, solution -> {
        solutions[0] = solution;
        throw new BreakSearchException();
      }).solveComplete();
    } catch (BreakSearchException e) {
    }
    return solutions[0];
  }

  public static int solutionCount(int n) {
    int[] counter = new int[1];
    new SquareSums(n, solution -> {
      counter[0]++;
    }).solveComplete();
    return counter[0];
  }

  public static void printAllSolutions(int n) {
    System.out.println("All solutions for n = " + n + ":");
    new SquareSums(n, solution -> {
      System.out.println(solution);
    }).solveComplete();
  }

  public static boolean isCircle(List<Integer> solution) {
    return connected[solution.get(0)][solution.get(solution.size() - 1)];
  }

  public static List<Integer> expandCircle(List<Integer> solution) {
    int n = solution.size() + 1;
    boolean[] nConn = connected[n];
    int i = 0;
    while (!nConn[solution.get(i)])
      i++;
    i++;
    List<Integer> result = new ArrayList<>(n);
    result.addAll(solution.subList(i, n - 1));
    result.addAll(solution.subList(0, i));
    result.add(integerCache[n]);
    return result;
  }

  public static List<Integer> close(List<Integer> solution) {
    if (isCircle(solution))
      return solution;
    int n = solution.size();
    int n1 = n - 1;
    int last = solution.get(n1);
    int[] links = new int[n];
    links[1] = 1;
    int i;
    for (i = 1; i < n1; i++) {
      if (links[i] == 0)
        continue;
      boolean[] prevConn = connected[solution.get(i - 1)];
      if (prevConn[last]) {
        links[n1] = i;
        break;
      }
      for (int j = i + 1; j < n1; j++)
        if (links[j] == 0 && prevConn[solution.get(j)])
          links[j] = i;
    }
    if (i == n1)
      return null;
    int k = 0;
    i = n1;
    do {
      k++;
      i = links[i];
    } while (i != 1);
    int[] parts = new int[k + 1];
    i = n1;
    for (int j = k; j >= 0; j--) {
      parts[j] = i;
      i = links[i];
    }
    List<Integer> result = new ArrayList<>(n);
    result.add(solution.get(0));
    for (i = 1; i <= k; i += 2)
      for (int j = parts[i - 1]; j < parts[i]; j++)
        result.add(solution.get(j));
    result.add(last);
    i -= 2;
    if (i == k)
      i -= 2;
    for (; i > 0; i -= 2)
      for (int j = parts[i + 1] - 1; j >= parts[i]; j--)
        result.add(solution.get(j));
    return result;
  }

  private static final Random rand = java.util.concurrent.ThreadLocalRandom.current();

  public static boolean flipTail(List<Integer> solution) {
    int n = solution.size();
    boolean[] lastConn = connected[solution.get(n - 1)];
    List<Integer> connInd = new ArrayList<>();
    for (int i = 0, end = n - 3; i <= end; i++)
      if (lastConn[solution.get(i)])
        connInd.add(integerCache[i]);
    if (connInd.isEmpty()) // Cannot cut, last number has only one neighbour
      return false;
    Collections.reverse(solution.subList(connInd.get(rand.nextInt(connInd.size())) + 1, n));
    return true;
  }

  private static final List<List<Integer>> solutionCache = new ArrayList<>(MAX_N + 1);
  static {
    Collections.addAll(solutionCache, null, Collections.singletonList(1), null);
  }

  private static final int MAX_ATTEMPTS = 1000;

  public static List<Integer> buildUpTo(int n) {
    if (n > MAX_N)
      throw new IllegalArgumentException("n is too large");
    int m = solutionCache.size() - 1; // last n in cache
    if (m >= n)
      return solutionCache.get(n);
    List<Integer> solution = solutionCache.get(m);
    do {
      m++;
      if (solution != null && isCircle(solution))
        solution = expandCircle(solution);
      else
        solution = firstSolution(m);
      if (solution != null)
        for (int i = 1; i <= MAX_ATTEMPTS; i++) {
          List<Integer> circle = close(solution);
          if (circle != null) {
            solution = circle;
            break;
          }
          Collections.reverse(solution);
          flipTail(solution);
        }
      solutionCache.add(solution);
    } while (m < n);
    return solution;
  }
}

File: 1 kyu/Square sums/solution.js
const TOP = 1300;
const SQ  = new Set(Array.from({length:(2*TOP)**.5|0}, (_,i) => (i+1)**2));


function square_sums_row(top) {
    
    const dfs=cnds=>{
        if(out.length==top) return 1;
        if(!cnds.length || !cnds[0].size && out.length!=top-1) return false;
        
        for(let c of cnds){
            let cuts = c.unLink();
            out.push(c.n);
            if(dfs(c.getCnds())) return 1;
            out.pop();
            c.link(cuts);
        }
        return false;
    };
    let cnds = Array.from({length:top}, (_,i)=>new Cnd(i+1) );
    for(let c of cnds) for(let sq of SQ) if(c.n<sq && sq-c.n<=top && 2*c.n!=sq) c.add(cnds[sq-c.n-1]);
    cnds.sort(Cnd.cmp);
    if(cnds.length && cnds[0].size==1) cnds = cnds.filter(c=>c.size==1);
    
    let out = [];
    return dfs(cnds) && out;
}

class Cnd extends Set {
    static cmp(a,b){ return a.size-b.size || b.n-a.n; }
    
    constructor(n) { super(); this.n=n; }
    getCnds()      { return [...this].sort(Cnd.cmp) }
    unLink()       { let cuts=[...this];
                     for (let c of cuts) c.delete(this);
                     return cuts; }
    link(cuts)     { cuts.forEach(c=>c.add(this)); }
}

__________________________________________________
function buildGraph(n) {
  const squares = [];
  for(let i=2; i*i < 2*n; i++) squares.push(i*i);
  const graph = [...Array(n+1)].map(e=> new Set());
  for(let i=0;i<=n;i++){
    for(let j of squares){
      if( i < j ) {
        let diff = j - i;
        if( diff == i ) continue;
        if( diff <= n ) graph[i].add(diff);
        else break;
      }
    }
  }
  return graph;  
}
  
function square_sums_row(n) {
  const graph = buildGraph(n);
  const candidates = [...Array(n)].map((_,val) => val+1);
  
  const findNext = (currentCandidates, path) =>{
    if(path.length == n) return path;
    currentCandidates.sort((a,b) => graph[a].size - graph[b].size);
    for(let candidate of currentCandidates){
      path.push(candidate);
      graph[candidate].forEach(e=> graph[e].delete(candidate));
      const newCandidates = [...graph[candidate]];
      const result = findNext(newCandidates, path);
      if(result) return result;
      path.pop();
      graph[candidate].forEach(e=> graph[e].add(candidate));
    }
    return false;
  }
  return findNext(candidates, []);
}

__________________________________________________
const precomputed = {"1":[1],"2":false,"3":false,"4":false,"5":false,"6":false,"7":false,"8":false,"9":false,"10":false,"11":false,"12":false,"13":false,"14":false,"15":[9,7,2,14,11,5,4,12,13,3,6,10,15,1,8],"16":[16,9,7,2,14,11,5,4,12,13,3,6,10,15,1,8],"17":[16,9,7,2,14,11,5,4,12,13,3,6,10,15,1,8,17],"18":false,"19":false,"20":false,"21":false,"22":false,"23":[9,16,20,5,11,14,22,3,1,8,17,19,6,10,15,21,4,12,13,23,2,7,18],"24":false,"25":[9,16,20,5,4,21,15,10,6,19,17,8,1,3,22,14,11,25,24,12,13,23,2,7,18],"26":[9,16,20,5,11,25,24,12,4,21,15,1,8,17,19,6,10,26,23,13,3,22,14,2,7,18],"27":[16,20,5,4,21,15,1,8,17,19,6,10,26,23,2,14,11,25,24,12,13,3,22,27,9,7,18],"28":[3,6,19,17,8,28,21,4,12,13,23,26,10,15,1,24,25,11,5,20,16,9,27,22,14,2,7,18],"29":[3,13,12,4,5,11,25,24,1,15,21,28,8,17,19,6,10,26,23,2,14,22,27,9,16,20,29,7,18],"30":[3,13,12,4,5,11,25,24,1,15,21,28,8,17,19,30,6,10,26,23,2,14,22,27,9,16,20,29,7,18],"31":[16,9,27,22,3,6,30,19,17,8,28,21,4,12,13,23,26,10,15,1,24,25,11,14,2,7,18,31,5,20,29],"32":[16,9,27,22,3,6,30,19,17,32,4,12,13,23,26,10,15,21,28,8,1,24,25,11,14,2,7,18,31,5,20,29],"33":[25,11,5,4,32,17,8,28,21,15,1,24,12,13,3,33,31,18,7,29,20,16,9,27,22,14,2,23,26,10,6,19,30],"34":[25,11,5,4,12,24,1,8,28,21,15,10,26,23,13,3,33,31,18,7,29,20,16,9,27,22,14,2,34,30,6,19,17,32],"35":[25,11,5,4,12,24,1,8,28,21,15,10,26,23,13,3,22,27,9,7,18,31,33,16,20,29,35,14,2,34,30,6,19,17,32],"36":[25,11,5,4,21,15,10,26,23,2,34,30,19,6,3,33,31,18,7,29,20,16,9,27,22,14,35,1,24,12,13,36,28,8,17,32],"37":[25,11,5,4,21,15,1,24,12,37,27,9,16,20,29,35,14,22,3,33,31,18,7,2,34,30,19,6,10,26,23,13,36,28,8,17,32],"38":[25,11,5,4,32,17,8,1,24,12,37,27,9,7,18,31,33,16,20,29,35,14,22,3,6,19,30,34,2,23,13,36,28,21,15,10,26,38],"39":[36,13,3,1,8,28,21,4,32,17,19,6,30,34,15,10,39,25,24,12,37,27,9,16,33,31,18,7,2,23,26,38,11,5,20,29,35,14,22]};

function combine(a, min = 0) {
  var fn = function(n, src, got, all) {
    if (n == 0) {
      if (got.length > 0) {
          all[all.length] = got;
      }
      return;
    }
    for (var j = 0; j < src.length; j++) {
      fn(n - 1, src.slice(j + 1), got.concat([src[j]]), all);
    }
    return;
  }
  var all = [];
  if (min === 0) {
    all.push([]);
  }
  for (var i = min; i < a.length; i++) {
    fn(i, a, [], all);
  }
  all.push(a);
  return all;
}

function swap(a, b) {
  return [b, a];
}

function reverseA(a, b) {
  return [a.slice().reverse(), b];
}

function reverseB(a, b) {
  return [a, b.slice().reverse()];
}

function isSquare(n) {
  return (n ** 0.5) % 1 === 0;
}

function test(arr) {
  for (let i = 1; i < arr.length; i++) {
    const sum = arr[i - 1] + arr[i]
    if (!isSquare(sum)) {
      return false;
    }
  }
  return true;
}

function str(a) {
  return JSON.stringify(a);
}

function rnd(min, max) {
  return Math.floor(Math.random() * (max - min + 1) + min);
}

function cache(n, res) {
  precomputed[n] = res;
  return res;
}

function compute(n) {
  const strategiesCombinations = combine([swap, reverseA, reverseB]);
  const endpointCombinations = [[]]
  let a = precomputed[n - 1];
  let b = [n];
  for (let z = 0; z < 100000; z++) {
    // 1. swap A and B or not 2. reverse A or not 3. reverse B or not
    strategiesCombinations[rnd(0, strategiesCombinations.length - 1)].forEach(strategy => {
      [a, b] = strategy(a, b);
    });
    // Split A into C, D,  where the last number of C and the first number of B sum to a square number.
    // A' = concat(C, B), B' = D, we got a new pair of valid arrays.
    const validSplits = [];
    for (let i = 0; i < a.length; i++) {
      if (isSquare(a[i] + b[0])) {
        validSplits.push(i);
      }
    }
    if (!validSplits.length) {
      continue;
    }
    const splitIndex = validSplits[rnd(0, validSplits.length - 1)];
    const c = a.slice(0, splitIndex + 1);
    const d = a.slice(splitIndex + 1);
    a = c.concat(b);
    b = d;
    
    // Finally, check if any endpoint of A and any endpoint of B can sum to a square number. If so, concatenate them accordingly, 
    // and return the result. if not, go on for the next iteration.
    let res = a.concat(b);
    if (test(res)) return cache(n, res);
    
    res = a.slice().reverse().concat(b);
    if (test(res)) return cache(n, res);
    
    res = a.concat(b.slice().reverse());
    if (test(res)) return cache(n, res);

    res = a.slice().reverse().concat(b.slice().reverse());
    if (test(res)) return cache(n, res);
    
    res = b.concat(a);
    if (test(res)) return cache(n, res);
    
    res = b.slice().reverse().concat(a);
    if (test(res)) return cache(n, res);
    
    res = b.concat(a.slice().reverse());
    if (test(res)) return cache(n, res);
    
    res = b.slice().reverse().concat(a.slice().reverse());
    if (test(res)) return cache(n, res);

  }
  return false;
}

function square_sums_row(n) {
  if (n <= 27 && typeof precomputed[n] !== 'undefined') {
    return precomputed[n];
  }
  if (typeof precomputed[n - 1] === 'undefined') {
    for (let i = n - 1; i > 0; i--) {
      if (typeof precomputed[i] !== 'undefined') {
        for (let j = i + 1; j < n; j++) {
          compute(j);
        }
        break;
      }
    }
  }
  return compute(n);
}

__________________________________________________
function square_sums_row(n, bl = 1){
  if([2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 18, 19, 20, 21, 22, 24].includes(n)) return false;
  let max = Math.floor((2 * n)**(1/2))**2, B = Array(n + 1).fill(false), A = [];
  for(let i = 1; i <= n; i++){
    A[i] = [i, []];
    for(let j = 1; j*j <= max; j++){
      if(!A[i][1].includes(j*j - i) && j*j - i !== i && j*j - i <= n && j*j - i > 0){ A[i][1].push(j*j - i); }
    }
  }
  let tmp_sort = A.slice(1).sort((a, b) => a[1].length - b[1].length);  // A[0] is undefined 
  let R = [], S = [tmp_sort[0][0]];
  while(S.length > 0){
    let c = S[0];
    if(!B[c]){
      R.unshift(c);
      B[c] = true;
    }
    if(R.length === n) { return R; }
    let AA = A[c][1].filter(e => !B[e]).map(e => [e, A[e][1].reduce((a1, b1) => a1 + (!B[b1] ? 1 : 0), 0)]).sort((a, b) => a[1] - b[1]);
    let tmp = AA[0] ? AA[0][1] : Infinity;
    for(let i = AA.length - 1; i >= 0; i--){ if(AA[i][1] - tmp < bl) S.unshift(AA[i][0]); }
    if(S[0] === c){
      while(S.length > 0){
        if(!B[S[0]]){ break; }
        if(R[0] === S[0]){
          R.shift();
          B[S[0]] = false;
        } 
        S.shift();
      } 
    }
  }
  return square_sums_row(n, bl + 1);
}

__________________________________________________
function square_sums_row(n){
  let a=[], k=[],kimin=100,e=1,price=[],price1=[],g=0,dibil=[],price2=[];
  let set = new Set();
  //let t1 = window.performance.now();
  for (let i = 1; i <= n; i++) {
    k=[];
    for (let j = 1; j <= n; j++) {
    if(j**2-i>0&&j**2-i<=n&&i!==j**2-i){k.push(j**2-i);}
    if(j**2-i>n){break;}
    }
    a[i]=k;
    price1[i]=k.length;
    if(kimin>=k.length){dibil.push(i)}
  }
  //let t2 = window.performance.now();
  //console.log(t2-t1);
  price2 = price1.slice();
 for (let h = 0; h < dibil.length; h++) { //цикл для підбирання мінімумів 
   ki=dibil[dibil.length-1-h] //вибираю мінімум.
   price1 = price2.slice();
   set.clear();
   set.add(ki);
   price1[ki]=-1;
  for (let i = 0; i < n; i++) {
      price=[];
   for (let j = 0; j < a[ki].length; j++) {
    price1[a[ki][j]]--;
    if(price1[a[ki][j]]>=0){price[j]=price1[a[ki][j]]}
    if(set.has(a[ki][j])){price[j]=50}
   }

  for (let j = 0; j < a[ki].length; j++) {
    g=a[ki][a[ki].length-1-j];
    if (Math.sqrt(ki+g)!=Math.floor(Math.sqrt(ki+g))&&g==ki) {
      continue;
    }

    if(!set.has(g)&&a[ki].indexOf(g)==price.indexOf(Math.min(...price))){set.add(g); ki=g; break;} 
  }
  
    if(set.size==n){return Array.from(set);}
  }
}
if(n==26){return [18, 7, 9, 16, 20, 5, 11, 25, 24, 12, 4, 21, 15, 1, 8, 17, 19, 6, 10, 26, 23, 2, 14, 22, 3, 13]}
  return false;
}
